### PR TITLE
Invalidate soft cache on change to entity relation

### PR DESCRIPTION
**Pull request by edmofro**
The soft cache was being correctly invalidated when an entity was modified, but some of the cached functions also depend on ancestor_descendant_relation. This adds invalidation when entity_relation changes, so that the soft cache doesn't end up with a stale version of the hierarchy.

### CODE REVIEW

File: packages/database/src/migrations/20201009035426-AddImmutableTableTrigger-modifies-schema.js [hunk 1]
```diff
@@ -31,7 +31,7 @@ exports.up = function (db) {
 
 exports.down = function (db) {
   return db.runSql(`
-    DROP FUNCTION IF EXISTS immutable_table();
+    DROP FUNCTION IF EXISTS immutable_table() CASCADE;
```
Comment (edmofro) [inline packages/database/src/migrations/20201009035426-AddImmutableTableTrigger-modifies-schema.js:34]: Unrelated, just found when I was trying to migrate down to replay the migrations

File: packages/database/src/runPostMigration.js [hunk 2]
```diff
@@ -21,14 +21,14 @@ const EXCLUDED_TABLES_FROM_TRIGGER_CREATION = [
 // attempted
 const IMMUTABLE_TABLES = ['ancestor_descendant_relation'];
 
-const SELECT_TABLES_WITHOUT_TRIGGERS = `
+const getSelectForTablesWithoutTriggers = triggerSuffix => `
```
Comment (edmofro) [inline packages/database/src/runPostMigration.js:24]: Changes in this file are unrelated, just found when I was trying to add both trigger types to ancestor_descendant_relation (ended up reverting that actual change)

File: packages/database/src/DatabaseModel.js [hunk 3]
```diff
@@ -18,17 +18,29 @@ export class DatabaseModel {
...
     if (this.database.isSingleton) {
+      // fully reset cache on any change to this model's records
       this.database.addChangeHandlerForCollection(this.DatabaseTypeClass.databaseType, () => {
-        this.cache = {}; // invalidate cache on any change
+        this.cache = {};
       });
+
+      // if this model has caching that depends on other models, also add invalidation for them
+      if (this.cacheDependencies) {
+        this.cacheDependencies.forEach(databaseType => {
```
Comment (kael89) [inline packages/database/src/DatabaseModel.js:35]: A minor suggestion, I see you used a getter for this property in `Entity.js`. Could we then define a base value for this property in `DatabaseModel.js`, i.e.

```js
  get cacheDependencies() {
    return [];
  }
```

or `this.cacheDependencies = [];`

In that case, we can avoid checking `if (this.cacheDependencies)`

File: packages/database/src/runPostMigration.js [hunk 4]
```diff
@@ -21,14 +21,14 @@ const EXCLUDED_TABLES_FROM_TRIGGER_CREATION = [
...
   SELECT t.table_name
     FROM information_schema.tables t
     LEFT JOIN information_schema.table_privileges privileges
       ON t.table_name = privileges.table_name
     WHERE NOT EXISTS (
       SELECT * from pg_trigger p
-      WHERE p.tgname LIKE concat(lower(t.table_name), '%', '_trigger')
+      WHERE p.tgname = concat(lower(t.table_name), '${triggerSuffix}')
```
Comment (kael89) [inline packages/database/src/runPostMigration.js:31]: I guess that '%' was not required here? (Probably not, just double checking

**Review by kael89 — APPROVED** [review]
This is an important change done in a simple and elegant way, nice! Approving with comments

File: packages/database/src/runPostMigration.js (hunk 4 above)
Comment (edmofro) [reply packages/database/src/runPostMigration.js:31]: Yeah no longer required. We were previously checking for `entity_trigger` and `entity_immutable_trigger` using a single `LIKE` comparison, with the `%` a wildcard for `_immutable`. Now we explicitly check for each

File: packages/database/src/DatabaseModel.js (hunk 3 above)
Comment (edmofro) [reply packages/database/src/DatabaseModel.js:35]: Good idea :-)

File: packages/database/src/runPostMigration.js (hunk 4 above)
Comment (kael89) [reply packages/database/src/runPostMigration.js:31]: I see - thanks for the explanation! 🙌
